gms/gossiper.cc

```diff
@@ -802,6 +758,87 @@ future<gossiper::endpoint_permit> gossiper::lock_endpoint(inet_address ep) {
     });
 }
 
+
+future<> gossiper::failure_detector_loop_for_node(gms::inet_address node, int64_t gossip_generation, uint64_t live_endpoints_version) {
+    auto last = gossiper::clk::now();
+    auto max_duration = std::chrono::milliseconds(_cfg.failure_detector_timeout_in_ms());
+    auto diff = std::chrono::milliseconds(0);
+    auto echo_interval = std::chrono::milliseconds(2000);
+    while (is_enabled()) {
+        bool failed = false;
+        try {
+            logger.debug("failure_detector_loop: Send echo to node {}, status = started", node);
+            co_await _messaging.send_gossip_echo(netw::msg_addr(node), gossip_generation, max_duration);
+            logger.debug("failure_detector_loop: Send echo to node {}, status = ok", node);
+        } catch (...) {
+            failed = true;
+            logger.warn("failure_detector_loop: Send echo to node {}, status = failed: {}", node, std::current_exception());
+        }
+        auto now = gossiper::clk::now();
+        diff = now - last;
+        if (!failed) {
+            last = now;
+        }
+        if (diff > max_duration) {
+            logger.info("failure_detector_loop: Mark node {} as DOWN", node);
+            co_await container().invoke_on(0, [node] (gms::gossiper& g) {
+                return seastar::async([node, &g] {
+                    g.convict(node);
+                });
+            });
+            co_return;
+        }
+
+        uint64_t current_live_endpoints_version = co_await container().invoke_on(0, [] (gms::gossiper& g) {
+            return g._live_endpoints_version;
+        });
```

Comment:
Ah, yes. It would be better to broadcast changes to _live_endpoints_version to all shards, than have each shard poll for changes, because _live_endpoints_version changes rarely.

Reply:
I want to avoid depending on the gossiper::run() where we copy values from zero shard to other shards.  We query _live_endpoints_version every two seconds. I think this is fine.

Reply:
> I want to avoid depending on the gossiper::run() where we copy values from zero shard to other shards. 

Why not? If it stalls, it will block the invoke_on() call too.

> We query _live_endpoints_version every two seconds. I think this is fine.

It's not a huge problem, but it violates the nice property of the patch that it avoids cross-shard calls in normal operation.

Reply:
It is pretty ugly to copy the replicate the data in the run loop. We moved some stuff out there. And I do not want to add new stuffs there. 

Anyway, in the latest version. I added a update_live_endpoints_version helper. Now, the nice property is preserved.